File: src/domains/shared/types/api-contracts.ts
// Shared API Contracts
// These types define the contracts between frontend and backend across all domains

// ============================================================================
// Core Data Types
// ============================================================================

export interface NetworkConfigRow {
  blueprint?: string;
  server_label?: string;
  is_external?: boolean;
  server_tags?: string;
  switch_tags?: string;
  link_group_ifname?: string;
  link_group_lag_mode?: string;
  link_group_ct_names?: string;
  link_group_tags?: string;
  link_speed?: string;
  server_ifname?: string;
  switch_label?: string;
  switch_ifname?: string;
  link_tags?: string;
  comment?: string;
}

// ============================================================================
// Processing and Results
// ============================================================================

export interface ProcessingResult {
  job_id: string;
  total_rows: number;
  processed_rows: number;
  failed_rows: number;
  success_rate: number;
  errors: ProcessingError[];
  status: ProcessingStatus;
}

export interface ProcessingError {
  row_index: number;
  error_message: string;
  error_type: ErrorType;
}

export enum ProcessingStatus {
  Pending = 'Pending',
  InProgress = 'InProgress',
  Completed = 'Completed',
  Failed = 'Failed',
  Cancelled = 'Cancelled',
}

export enum ErrorType {
  ValidationError = 'ValidationError',
  NetworkError = 'NetworkError',
  DataError = 'DataError',
  SystemError = 'SystemError',
}

// ============================================================================
// Validation Types
// ============================================================================

export interface ValidationError {
  row_index: number;
  field: string;
  message: string;
}

export interface DataValidationResult {
  valid_rows: NetworkConfigRow[];
  invalid_rows: NetworkConfigRow[];
  errors: ValidationError[];
  duplicates_removed: number;
}

// ============================================================================
// Configuration Types
// ============================================================================

export interface ApstraConfig {
  host: string;
  port: number;
  username: string;
  password: string;
  blueprint_name: string;
  use_ssl?: boolean;
  verify_ssl?: boolean;
  timeout?: number;
}

export interface ConversionMap {
  header_row?: number;
  mappings: Record<string, string>;
}

// ============================================================================
// API Response Types
// ============================================================================

export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface AuthResult {
  success: boolean;
  token?: string;
  expires_at?: string;
  error?: string;
}

export interface SystemSearchResult {
  id: string;
  label: string;
  hostname?: string;
  system_type: string;
  blueprint_id?: string;
}

export interface QueryResult {
  data: any;
  errors?: string[];
  execution_time?: number;
}

// ============================================================================
// File Processing Types
// ============================================================================

export interface ExcelFileInfo {
  file_path: string;
  sheet_names: string[];
  total_sheets: number;
}

export interface ExcelSheetData {
  sheet_name: string;
  headers: string[];
  rows: any[][];
  total_rows: number;
}

// ============================================================================
// Tauri Command Contracts
// ============================================================================

export interface TauriCommands {
  // Excel Domain Commands
  'upload_excel_file': (filePath: string) => Promise<string[]>;
  'parse_excel_sheet': (filePath: string, sheetName: string) => Promise<NetworkConfigRow[]>;
  
  // Apstra Domain Commands
  'apstra_login': (config: ApstraConfig) => Promise<AuthResult>;
  'apstra_search_systems': (query: string) => Promise<SystemSearchResult[]>;
  'apstra_execute_query': (query: string) => Promise<QueryResult>;
  'test_apstra_connection': (config: ApstraConfig) => Promise<ApiResponse<boolean>>;
  
  // Conversion Domain Commands
  'convert_headers_enhanced': (headers: string[], map: ConversionMap) => Promise<ApiResponse<any>>;
  'load_enhanced_conversion_map': () => Promise<ApiResponse<ConversionMap>>;
  'save_enhanced_conversion_map': (map: ConversionMap) => Promise<ApiResponse<boolean>>;
  
  // Provisioning Domain Commands
  'process_import_generic_system': (data: NetworkConfigRow[], config: ApstraConfig) => Promise<ProcessingResult>;
  'get_processing_progress': (jobId: string) => Promise<ProcessingResult>;
  
  // Shared Commands
  'send_backend_log': (level: string, message: string) => Promise<void>;
  'cleanup_temp_file': (filePath: string) => Promise<boolean>;
}

// ============================================================================
// Domain Event Types
// ============================================================================

export interface DomainEvent {
  domain: string;
  event_type: string;
  payload: any;
  timestamp: string;
}

export interface ProcessingProgressEvent extends DomainEvent {
  domain: 'provisioning';
  event_type: 'processing_progress';
  payload: {
    job_id: string;
    progress: number;
    status: ProcessingStatus;
    current_row?: number;
    total_rows?: number;
  };
}

export interface AuthStatusEvent extends DomainEvent {
  domain: 'apstra';
  event_type: 'auth_status_changed';
  payload: {
    authenticated: boolean;
    expires_at?: string;
  };
}

// ============================================================================
// Error Handling Types
// ============================================================================

export interface DomainError {
  domain: string;
  error_type: string;
  message: string;
  details?: any;
  timestamp: string;
}

export interface ExcelError extends DomainError {
  domain: 'excel';
  error_type: 'file_not_found' | 'invalid_sheet' | 'parse_error';
}

export interface ApstraError extends DomainError {
  domain: 'apstra';
  error_type: 'auth_failed' | 'api_request_failed' | 'invalid_config';
}

export interface ConversionError extends DomainError {
  domain: 'conversion';
  error_type: 'mapping_error' | 'validation_error' | 'transformation_error';
}

export interface ProvisioningError extends DomainError {
  domain: 'provisioning';
  error_type: 'processing_error' | 'network_error' | 'data_error';
}

File: src-tauri/src/domains/shared/models/api_contracts.rs
// Shared API Contracts
// These types define the contracts between frontend and backend across all domains

use serde::{Deserialize, Serialize};

// ============================================================================
// Core Data Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct NetworkConfigRow {
    pub blueprint: Option<String>,
    pub server_label: Option<String>,
    pub is_external: Option<bool>,
    pub server_tags: Option<String>,
    pub switch_tags: Option<String>,
    pub link_group_ifname: Option<String>,
    pub link_group_lag_mode: Option<String>,
    pub link_group_ct_names: Option<String>,
    pub link_group_tags: Option<String>,
    pub link_speed: Option<String>,
    pub server_ifname: Option<String>,
    pub switch_label: Option<String>,
    pub switch_ifname: Option<String>,
    pub link_tags: Option<String>,
    pub comment: Option<String>,
}

// ============================================================================
// Processing and Results
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessingResult {
    pub job_id: String,
    pub total_rows: usize,
    pub processed_rows: usize,
    pub failed_rows: usize,
    pub success_rate: f64,
    pub errors: Vec<ProcessingError>,
    pub status: ProcessingStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessingError {
    pub row_index: usize,
    pub error_message: String,
    pub error_type: ErrorType,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ProcessingStatus {
    Pending,
    InProgress,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ErrorType {
    ValidationError,
    NetworkError,
    DataError,
    SystemError,
}

// ============================================================================
// Validation Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationError {
    pub row_index: usize,
    pub field: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataValidationResult {
    pub valid_rows: Vec<NetworkConfigRow>,
    pub invalid_rows: Vec<NetworkConfigRow>,
    pub errors: Vec<ValidationError>,
    pub duplicates_removed: usize,
}

// ============================================================================
// Configuration Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApstraConfig {
    pub host: String,
    pub port: u16,
    pub username: String,
    pub password: String,
    pub blueprint_name: String,
    pub use_ssl: Option<bool>,
    pub verify_ssl: Option<bool>,
    pub timeout: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversionMap {
    pub header_row: Option<usize>,
    pub mappings: std::collections::HashMap<String, String>,
}

// ============================================================================
// API Response Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiResponse<T> {
    pub success: bool,
    pub data: Option<T>,
    pub error: Option<String>,
    pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthResult {
    pub success: bool,
    pub token: Option<String>,
    pub expires_at: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemSearchResult {
    pub id: String,
    pub label: String,
    pub hostname: Option<String>,
    pub system_type: String,
    pub blueprint_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryResult {
    pub data: serde_json::Value,
    pub errors: Option<Vec<String>>,
    pub execution_time: Option<u64>,
}

// ============================================================================
// File Processing Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExcelFileInfo {
    pub file_path: String,
    pub sheet_names: Vec<String>,
    pub total_sheets: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExcelSheetData {
    pub sheet_name: String,
    pub headers: Vec<String>,
    pub rows: Vec<Vec<serde_json::Value>>,
    pub total_rows: usize,
}

// ============================================================================
// Domain Event Types
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DomainEvent {
    pub domain: String,
    pub event_type: String,
    pub payload: serde_json::Value,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessingProgressEvent {
    pub domain: String, // Should be "provisioning"
    pub event_type: String, // Should be "processing_progress"
    pub payload: ProcessingProgressPayload,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessingProgressPayload {
    pub job_id: String,
    pub progress: f64,
    pub status: ProcessingStatus,
    pub current_row: Option<usize>,
    pub total_rows: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthStatusEvent {
    pub domain: String, // Should be "apstra"
    pub event_type: String, // Should be "auth_status_changed"
    pub payload: AuthStatusPayload,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthStatusPayload {
    pub authenticated: bool,
    pub expires_at: Option<String>,
}

// ============================================================================
// Default Implementations
// ============================================================================

impl Default for ProcessingResult {
    fn default() -> Self {
        Self {
            job_id: String::new(),
            total_rows: 0,
            processed_rows: 0,
            failed_rows: 0,
            success_rate: 0.0,
            errors: Vec::new(),
            status: ProcessingStatus::Pending,
        }
    }
}

impl Default for ConversionMap {
    fn default() -> Self {
        Self {
            header_row: Some(1),
            mappings: std::collections::HashMap::new(),
        }
    }
}

impl<T> Default for ApiResponse<T> {
    fn default() -> Self {
        Self {
            success: false,
            data: None,
            error: None,
            message: None,
        }
    }
}

impl<T> ApiResponse<T> {
    pub fn success(data: T) -> Self {
        Self {
            success: true,
            data: Some(data),
            error: None,
            message: None,
        }
    }

    pub fn error(error: String) -> Self {
        Self {
            success: false,
            data: None,
            error: Some(error),
            message: None,
        }
    }

    pub fn with_message(mut self, message: String) -> Self {
        self.message = Some(message);
        self
    }
}